pkg/reconciler/topic/topic.go

```diff
@@ -0,0 +1,418 @@
+				"VerifyFailed",
+				"Failed to verify Topic: %q",
+				msg)
+			return err
+		}
+	default:
+		logger.Error("Unknown propagation policy.", zap.Any("propagationPolicy", topic.Spec.PropagationPolicy))
+		return nil
+	}
+
+	publisher, err := c.createOrUpdatePublisher(ctx, topic)
+	if err != nil {
+		logger.Error("Unable to create the publisher", zap.Error(err))
+		return err
+	}
+	topic.Status.PropagatePublisherStatus(publisher.Status.GetCondition(apis.ConditionReady))
+
+	if publisher.Status.IsReady() {
+		topic.Status.SetAddress(publisher.Status.Address.URL)
+	}
+	return nil
+}
+
+func (c *Reconciler) updateStatus(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, error) {
+	topic, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(topic.Status, desired.Status) {
+		return topic, nil
+	}
+	becomesReady := desired.Status.IsReady() && !topic.Status.IsReady()
+	// Don't modify the informers copy.
+	existing := topic.DeepCopy()
+	existing.Status = desired.Status
+
+	ch, err := c.RunClientSet.PubsubV1alpha1().Topics(desired.Namespace).UpdateStatus(existing)
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Topic %q became ready after %v", topic.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Topic", topic.Namespace, topic.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Topic, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) updateFinalizers(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, bool, error) {
+	source, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Don't modify the informers copy.
+	existing := source.DeepCopy()
+
+	var finalizers []string
+
+	// If there's nothing to update, just return.
+	exisitingFinalizers := sets.NewString(existing.Finalizers...)
+	desiredFinalizers := sets.NewString(desired.Finalizers...)
+
+	if desiredFinalizers.Has(finalizerName) {
+		if exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Add the finalizer.
+		finalizers = append(existing.Finalizers, finalizerName)
+	} else {
+		if !exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Remove the finalizer.
+		exisitingFinalizers.Delete(finalizerName)
+		finalizers = exisitingFinalizers.List()
+	}
+
+	mergePatch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers":      finalizers,
+			"resourceVersion": existing.ResourceVersion,
+		},
+	}
+
+	patch, err := json.Marshal(mergePatch)
+	if err != nil {
+		return desired, false, err
+	}
+
+	update, err := c.RunClientSet.PubsubV1alpha1().Topics(existing.Namespace).Patch(existing.Name, types.MergePatchType, patch)
```

Comment:
Why use a patch instead of an update here?

Reply:
we are only trying to change the finalizers

Reply:
Update semantics are the same - either way, ResourceVersion will cause the op to fail if it changes in the meantime.

Reply:
@n3wscott says: This is patch to defend against bad controller practice of changing the spec in the reconciler.